go/inst/instance_topology_dao.go

```diff
@@ -579,14 +581,172 @@ func StartReplicationUntilMasterCoordinates(instanceKey *InstanceKey, masterCoor
+// If the flags are both turned off, the legacy behavior kicks in: If SemiSyncPriority > 0 and the instance is promotable (not "must_not"),
+// semi-sync is enabled.
+func MaybeEnableSemiSyncReplica(replicaInstance *Instance) (*Instance, error) {
+	// Backwards compatible logic: Enable semi-sync if SemiSyncPriority > 0 (formerly SemiSyncEnforced)
+	// Note that this logic NEVER enables semi-sync if the promotion rule is "must_not".
+	if !config.Config.EnforceExactSemiSyncReplicas && !config.Config.RecoverLockedSemiSyncMaster {
+		if replicaInstance.SemiSyncPriority > 0 {
+			enable := replicaInstance.PromotionRule != MustNotPromoteRule // Send ACK only from promotable instances
+			log.Infof("semi-sync: %+v: setting rpl_semi_sync_slave_enabled = %t (legacy behavior)", &replicaInstance.Key, enable)
+			return SetSemiSyncReplica(&replicaInstance.Key, enable)
+		}
+		return replicaInstance, nil
+	}
+
+	// New logic: If EnforceExactSemiSyncReplicas or RecoverLockedSemiSyncMaster are set, we enable semi-sync only if the
+	// given replica instance is in the list of replicas to have semi-sync enabled (according to the priority).
+	masterInstance, err := ReadTopologyInstance(&replicaInstance.MasterKey)
+	if err != nil {
+		return replicaInstance, log.Errore(err)
+	}
+	replicas, err := ReadReplicaInstances(&replicaInstance.MasterKey)
+	if err != nil {
+		return replicaInstance, log.Errore(err)
+	}
+
+	possibleSemiSyncReplicas, asyncReplicas, excludedReplicas := ClassifyAndPrioritizeReplicas(replicas, &replicaInstance.Key)
+	actions := DetermineSemiSyncReplicaActions(possibleSemiSyncReplicas, asyncReplicas, masterInstance.SemiSyncMasterWaitForReplicaCount, masterInstance.SemiSyncMasterClients, config.Config.EnforceExactSemiSyncReplicas)
+
+	log.Debugf("semi-sync: %+v: determining whether to enable rpl_semi_sync_slave_enabled", replicaInstance.Key)
+	log.Debugf("semi-sync: master = %+v, master semi-sync wait count = %d, master semi-sync replica count = %d", replicaInstance.MasterKey, masterInstance.SemiSyncMasterWaitForReplicaCount, masterInstance.SemiSyncMasterClients)
+	LogSemiSyncReplicaAnalysis(possibleSemiSyncReplicas, asyncReplicas, excludedReplicas, actions, func(s string, a ...interface{}) { log.Debugf(s, a...) })
+
+	// TODO This is a little odd. We sometimes get actions for replicas that are not us. If we do not take this action
+	// TODO then we'll absolutely have a MasterWithTooManySemiSyncReplicas event following this recovery. We could prevent this
+	// TODO by just executing all actions, but we're in the scope of `replicaInstance` here, so..... idk
+	for replica, enable := range actions {
+		if replica.Key.Equals(&replicaInstance.Key) {
+			log.Infof("semi-sync: %s: setting rpl_semi_sync_slave_enabled: %t", &replicaInstance.Key, enable)
+			return SetSemiSyncReplica(&replicaInstance.Key, enable)
+		}
+	}
```

Comment:
Feels like this code and `recoverSemiSyncReplicas` have a lot in common. I don't feel strongly, but I think there's a significant amount of logic we can extract to an external function.

Reply:
Your call. I don't feel strongly.